pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AntlrBaseNode.java

```diff
@@ -0,0 +1,93 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.ast;
+
+import java.util.Iterator;
+
+import org.antlr.v4.runtime.ParserRuleContext;
+
+import net.sourceforge.pmd.lang.ast.xpath.Attribute;
+import net.sourceforge.pmd.lang.ast.xpath.AttributeAxisIterator;
+import net.sourceforge.pmd.lang.dfa.DataFlowNode;
+
+public class AntlrBaseNode extends ParserRuleContext implements AntlrNode {
+
+    // TODO: what should we do with parent? how do we handle data flows in this scenario? it's ok to ignore
+    // TODO: our parent data flow in case we don't have one?
+    // protected Node parent;
+
+    private DataFlowNode dataFlowNode;
+    private Object userData;
+
+    /**
+     * Constructor required by {@link ParserRuleContext}
+     */
+    @SuppressWarnings("unused")
+    public AntlrBaseNode() {
+        // Nothing to be done
+    }
+
+    /**
+     * Constructor required by {@link ParserRuleContext}
+     *
+     * @param parent The parent
+     * @param invokingStateNumber the invokingState defined by {@link org.antlr.v4.runtime.RuleContext} parent
+     */
+    @SuppressWarnings("unused")
+    public AntlrBaseNode(final ParserRuleContext parent, final int invokingStateNumber) {
+        super(parent, invokingStateNumber);
+    }
+
+    @Override
+    public int getBeginLine() {
+        return start.getLine(); // This goes from 1 to n
+    }
+
+    @Override
+    public int getEndLine() {
+        return stop.getLine(); // This goes from 1 to n
+    }
+
+    @Override
+    public int getBeginColumn() {
+        return start.getCharPositionInLine(); // This goes from 0 to (n - 1)
+    }
+
+    @Override
+    public int getEndColumn() {
+        return stop.getCharPositionInLine(); // This goes from 0 to (n - 1)
+    }
+
+    @Override
+    public DataFlowNode getDataFlowNode() {
+        return dataFlowNode;
+    }
+
+    @Override
+    public void setDataFlowNode(final DataFlowNode dataFlowNode) {
+        this.dataFlowNode = dataFlowNode;
+    }
+
+    @Override
+    public Object getUserData() {
+        return userData;
+    }
+
+    @Override
+    public void setUserData(final Object userData) {
+        this.userData = userData;
+    }
+
+    // TODO: should we make it abstract due to the comment in AbstractNode ?
+    @Override
+    public String getXPathNodeName() {
+        return toString();
```

Comment:
given this is a base node for all antlr classes, you can probably provide a default implementation that will return the node type. If you can't, leave it out. You probably can't, as the best implementation would be `StringUtils.capitalize(LanguageParser.ruleNames[node.getRuleIndex()])` or something similar and you don't know here which `LanguageParser` you need

Reply:
Ok, we will do some research about this.

Reply:
Finally we are going to leave this implementation as it is.

Reply:
What is it currently producing for Antlr nodes? is it a sane default? Or are we better off leaving this as abstract?